lib/src/connection_checker.dart

```diff
@@ -0,0 +1,164 @@
+import 'dart:async';
+import 'dart:io';
+
+import 'package:connecteo/src/connection_type.dart';
+import 'package:connecteo/src/connection_type_mapper.dart';
+import 'package:connecteo/src/host_reachability_checker.dart';
+import 'package:connectivity_plus/connectivity_plus.dart';
+import 'package:rxdart/rxdart.dart';
+
+const _connectionRequestInterval = Duration(seconds: 3);
+
+const _connectionFailureAttempts = 4;
+
+class ConnectionChecker {
+  ConnectionChecker._({
+    List<InternetAddress>? checkAddresses,
+    Duration? checkOverDnsTimeout,
+    String? baseUrlLookupAddress,
+  })  : _checkAddresses = checkAddresses,
+        _checkOverDnsTimeout = checkOverDnsTimeout,
+        _baseUrlLookupAddress = baseUrlLookupAddress,
+        _connectivity = Connectivity(),
+        _connectionTypeMapper = ConnectionTypeMapper(),
+        _hostReachabilityChecker = HostReachabilityChecker(),
+        _initialConnectedValue = Completer(),
+        _connectionTypeStreamController =
+            BehaviorSubject.seeded(ConnectionType.unknown) {
+    _setupInitialConnection().then((_) => _setupConnectivityListener());
+  }
+
+  factory ConnectionChecker({
+    List<InternetAddress>? customCheckAddresses,
+    Duration? customCheckTimeout,
+    String? baseUrlLookupAddress,
+  }) {
+    return _singleton ??= ConnectionChecker._(
+      checkAddresses: customCheckAddresses,
+      checkOverDnsTimeout: customCheckTimeout,
+      baseUrlLookupAddress: baseUrlLookupAddress,
+    );
+  }
+
+  static ConnectionChecker? _singleton;
+
+  final Connectivity _connectivity;
+  final ConnectionTypeMapper _connectionTypeMapper;
+
+  final BehaviorSubject<ConnectionType> _connectionTypeStreamController;
+  late final StreamSubscription<ConnectionType> _connectionTypeSubscription;
+
+  final List<InternetAddress>? _checkAddresses;
+  final Duration? _checkOverDnsTimeout;
+  final String? _baseUrlLookupAddress;
+  final HostReachabilityChecker _hostReachabilityChecker;
+
+  final Completer<bool> _initialConnectedValue;
+
+  Future<void> dispose() async {
+    await _connectionTypeSubscription.cancel();
+    await _connectionTypeStreamController.close();
```

Comment:
Maybe we can somehow avoid using thinks that need disposing/closing like controllers and subscriptions?